consensus/hotstuff/signature/weighted_signature_aggregator.go

```diff
@@ -1,39 +1,141 @@
+		return nil, fmt.Errorf("new signature aggregator failed: %w", err)
+	}
+
+	// build the weighted aggregator
+	weightedAgg := &WeightedSignatureAggregator{
+		SignatureAggregatorSameMessage: agg,
+		signers:                        signers,
+	}
 
-// Verify returns true if and only if the signature is valid.
-// It expects that correct type of signature is passed. Only SigTypeStaking is supported
+	// build the internal maps for a faster look-up
+	for i, id := range signers {
+		weightedAgg.idToIndex[id.NodeID] = i
+		weightedAgg.idToWeights[id.NodeID] = id.Stake
+	}
+	return weightedAgg, nil
+}
+
+// Verify verifies the signature under the stored public and message.
 func (s *WeightedSignatureAggregator) Verify(signerID flow.Identifier, sig crypto.Signature) (bool, error) {
-	panic("to be implemented")
+	index, ok := s.idToIndex[signerID]
+	if !ok {
+		return false, fmt.Errorf("couldn't find signerID %s in the index map", signerID)
+	}
+	return s.SignatureAggregatorSameMessage.Verify(index, sig)
 }
 
-// TrustedAdd adds an already verified signature, and look up the weight for the given signer,
-// and add it to the total weight, and returns the total weight that have been collected.
-// return (1000, nil) means the signature has been added, and 1000 weight has been collected in total.
-// return (1000, nil) means the signature is a duplication and 1000 weight has been collected in total.
-func (s *WeightedSignatureAggregator) TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (totalWeight uint64, exception error) {
-	panic("to be implemented")
+// TrustedAdd adds a signature to the internal set of signatures.
+//
+// It adds the signer's weight to the total collected weight and returns the total weight regardless
+// of the returned error.
+// The function errors if a signature from the signerID was already collected.
+// The function is thread-safe
+func (s *WeightedSignatureAggregator) TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (uint64, error) {
+	// get the total weight safely
+	collectedWeight := s.TotalWeight()
+
+	// get the index
+	index, ok := s.idToIndex[signerID]
+	if !ok {
+		return collectedWeight, fmt.Errorf("couldn't find signerID %s in the index map", signerID)
+	}
+	// get the weight
+	weight, ok := s.idToWeights[signerID]
+	if !ok {
+		return collectedWeight, fmt.Errorf("couldn't find signerID %s in the weight map", signerID)
+	}
+
+	// atomically update the signatures pool and the total weight
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// This is a sanity check because the upper layer should have already checked for double-voters.
+	_, ok = s.collectedIDs[signerID]
+	if ok {
+		return collectedWeight, fmt.Errorf("SigneID %s was already added", signerID)
+	}
+
+	err := s.SignatureAggregatorSameMessage.TrustedAdd(index, sig)
+	if err != nil {
+		return collectedWeight, fmt.Errorf("Trusted add has failed: %w", err)
+	}
+
+	s.collectedIDs[signerID] = struct{}{}
+	collectedWeight += weight
```

Comment:
You may be observing not latest value. You have no guarantees that `collectedWeight` didn't update prior locking the mutex. 
I would suggest checking next code to:
```
s.totalWeight += weight
```
instead of assigning to `s.totalWeight`

Reply:
agree with @durkmurder .

Reply:
As mentioned in the PR description, I didn't target `WeightedSignatureAggregator` in this PR. I [commented](https://github.com/onflow/flow-go/pull/1292/commits/43d934cd788215d3075fe9971c3c69b8f8277a0b) the code so we can focus on `SignatureAggregatorSameMessage` here. Sorry for the confusion. (I'll still consider your comments for the next PR)